File: adventofcode/2020/day3/day3.rb
# frozen_string_literal: true

def read_input
  ARGF.readlines
end

TREE = '#'
OPEN = '.'
SLOPES = [3, 1].freeze

def tree_count(terrain, slope)
  tree_count = 0
  row = 0
  col = 0
  row_max = terrain.length - 1
  col_max = terrain[0].length - 1
  while row < row_max
    if col + slope[0] > col_max # wrap around
      col = slope[0] - (col_max - col) - 1
    else
      col += slope[0]
    end
    row += slope[1]
    break if row > row_max

    tree_count += 1 if terrain[row][col] == TREE
  end
  tree_count
end

def part1(terrain)
  tree_count(terrain, SLOPES)
end

ALL_SLOPES = [
  [1, 1],
  [3, 1],
  [5, 1],
  [7, 1],
  [1, 2]
].freeze

def part2(terrain)
  ALL_SLOPES.map do |slope|
    tree_count(terrain, slope)
  end.inject(:*)
end

terrain = read_input.map { |r| r.scan(/./) }
puts part1(terrain)
puts part2(terrain)

File: adventofcode/2020/day3/day3.hs
tree = '#'
open = '.'

data Pos = Pos Int Int
data Slope = Slope Int Int

type Terrain = [String]

atPos :: Pos -> Terrain -> Char
atPos (Pos x y) ts = chr
  where row = ts !! y
        chr = (!! x) $ concat $ repeat row

movePos :: Pos -> Slope -> Pos
movePos (Pos x y) (Slope a b) = Pos (x + a) (y + b)

isTree :: Pos -> Terrain -> Bool
isTree = ((== tree) .) . atPos

isOpen :: Pos -> Terrain -> Bool
isOpen = (.) (== open) . atPos

treeCount :: Pos -> Slope -> Terrain -> Int
treeCount p@(Pos _ y) s ts
  | y >= length ts = 0
  | isTree p ts = 1 + treeCount np s ts
  | otherwise = treeCount np s ts
  where np = movePos p s

part1 :: Slope -> Terrain -> Int
part1 = treeCount (Pos 0 0)

part2 :: [Slope] -> Terrain -> Int
part2 ss ts = foldl (*) 1 $ map (\s -> part1 s ts) ss

solve :: Terrain -> String
solve ts = unlines $ map show $ [ part1 (Slope 3 1) ts
                                , part2 [ Slope 1 1
                                        , Slope 3 1
                                        , Slope 5 1
                                        , Slope 7 1
                                        , Slope 1 2
                                        ] ts
                                ]

main :: IO ()
main = interact $ solve . lines
